### PR TITLE
OSDOCS#6831: Recovering from etcd minority failure

### PR DESCRIPTION
**Pull request by obrown1205**
Version(s):
4.14+

Issue:
https://issues.redhat.com/browse/OSDOCS-7613

Link to docs preview:
https://67790--docspreview.netlify.app/openshift-enterprise/latest/hosted_control_planes/hcp-backup-restore-dr

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

**Comment by csrwng** [comment]
@obrown1205 I believe these instructions are a bit outdated. The latest version of these instructions is here: https://hypershift-docs.netlify.app/how-to/etcd-recovery/#recovery-from-quorum-loss
( source here: https://github.com/openshift/hypershift/blob/main/docs/content/how-to/etcd-recovery.md?plain=1#L58-L239)

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            sleep 1000
+          done
+      volumes:
+      - name: data
+        persistentVolumeClaim:
+          claimName: data-etcd-0
+EOF
+----
++
+.... Clear previous data and restore the snapshot
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:178]: delete (redundant)

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
++
+..... Check the status of the `etcd-data` pod by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:180]: Fix the indentation and clarify that this is to wait for the pod to be running

```
.. Check the status of the `etcd-data` pod and wait for it to be in a running state by entering the following command:
```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc get -n ${CONTROL_PLANE_NAMESPACE} pods -l app=etcd-data
+----
+
+..... Get the name of the `etcd-data` pod by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:187]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ DATA_POD=$(oc get -n ${CONTROL_PLANE_NAMESPACE} pods --no-headers -l app=etcd-data -o name | cut -d/ -f2)
+----
+
+..... Copy an etcd snapshot into the pod by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:194]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc cp /tmp/etcd.snapshot.db ${CONTROL_PLANE_NAMESPACE}/${DATA_POD}:/var/lib/restored.snap.db
+----
++
+..... Remove old data from the `etcd-data` pod by entering the following commands:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:201]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- rm -rf /var/lib/data
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- mkdir -p /var/lib/data
+----
++
+..... Restore an etcd snapshot by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:209]: 2 dots

s/Restore an etcd snapshot/Restore the etcd snapshot/

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+----
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- etcdutl snapshot restore /var/lib/restored.snap.db \
+     --data-dir=/var/lib/data --skip-hash-check \
+     --name etcd-0 \
+     --initial-cluster-token=etcd-cluster \
+     --initial-cluster etcd-0=https://etcd-0.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380,etcd-1=https://etcd-1.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380,etcd-2=https://etcd-2.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380 \
+     --initial-advertise-peer-urls https://etcd-0.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380
+----
++
+..... Remove an etcd snapshot from the `etcd-0` directory by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:221]:
```
.. Clean up the temporary snapshot from the pod by entering the following command:
```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- rm /var/lib/restored.snap.db
+----
++
+.... Delete data access deployment by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:228]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+`oc delete -n ${CONTROL_PLANE_NAMESPACE} deployment/etcd-data`
+----
++
+.... Scale up an etcd cluster by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:235]:
```
.. Scale up the etcd cluster by entering the following command:
```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc scale -n ${CONTROL_PLANE_NAMESPACE} statefulset/etcd --replicas=3
+----
++
+... Wait for the etcd member pods to return and report as available by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:242]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc get -n ${CONTROL_PLANE_NAMESPACE} pods -l app=etcd -w
+----
++
+.... Scale up all etcd-writer deployments by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:249]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
++
+[source,terminal]
+----
+$ oc scale deployment -n ${HOSTED_CLUSTER_NAMESPACE} --replicas=3 kube-apiserver openshift-apiserver openshift-oauth-apiserver
+----
++
+.... Start reconciliation of the hosted cluster by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:256]:
```
.. Restore reconciliation of the hosted cluster by entering the following command:
```

**Review by csrwng — COMMENTED** [review]
Some fixes to wording an indentation

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+----
++
+... Make a local copy of the snapshot by entering the following command:
++
+[source,terminal]
+----
+`oc cp -c etcd ${CONTROL_PLANE_NAMESPACE}/${ETCD_POD}:/var/lib/snapshot.db /tmp/etcd.snapshot.db`
+----
++
+... Make a copy of the snapshot database from etcd persistent storage.
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:86]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
++
+... List etcd pods by entering the following command:
++
+[source,terminal]
+----
+`oc get -n ${CONTROL_PLANE_NAMESPACE} pods -l app=etcd` 
+----
++
+... Find a pod that is running and set its name as the value of `ETCD_POD ETCD_POD=etcd-0`, then copy its snapshot database by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:95]: change: set its name as the value of `ETCD_POD ETCD_POD=etcd-0`
to: set its name as the value of `ETCD_POD`:  `ETCD_POD=etcd-0`

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+----
++
+. Scale down the etcd statefulset by entering the following command:
++
+[source,terminal]
+----
+`oc scale -n ${CONTROL_PLANE_NAMESPACE} statefulset/etcd --replicas=0`
+----
++
+.. Delete volumes for second and third members by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:109]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+`oc delete -n ${CONTROL_PLANE_NAMESPACE} pvc/data-etcd-1 pvc/data-etcd-2`
+----
++
+. Create a pod to access the first etcd member's data:
++
+.... Get the etcd image by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:118]: 2 dots

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+----
+
+.. Get the name of the `etcd-data` pod by entering the following command:
++
+[source,terminal]
+----
+$ DATA_POD=$(oc get -n ${CONTROL_PLANE_NAMESPACE} pods --no-headers -l app=etcd-data -o name | cut -d/ -f2)
+----
+
+.. Copy an etcd snapshot into the pod by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:179]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+$ oc cp /tmp/etcd.snapshot.db ${CONTROL_PLANE_NAMESPACE}/${DATA_POD}:/var/lib/restored.snap.db
+----
++
+.. Remove old data from the `etcd-data` pod by entering the following commands:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:186]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- rm -rf /var/lib/data
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- mkdir -p /var/lib/data
+----
++
+.. Restore the etcd snapshot by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:194]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+----
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- etcdutl snapshot restore /var/lib/restored.snap.db \
+     --data-dir=/var/lib/data --skip-hash-check \
+     --name etcd-0 \
+     --initial-cluster-token=etcd-cluster \
+     --initial-cluster etcd-0=https://etcd-0.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380,etcd-1=https://etcd-1.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380,etcd-2=https://etcd-2.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380 \
+     --initial-advertise-peer-urls https://etcd-0.etcd-discovery.${CONTROL_PLANE_NAMESPACE}.svc:2380
+----
++
+.. Remove the temporary etcd snapshot from the pod by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:206]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+$ oc exec -n ${CONTROL_PLANE_NAMESPACE} ${DATA_POD} -- rm /var/lib/restored.snap.db
+----
++
+.. Delete data access deployment by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:213]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+`oc delete -n ${CONTROL_PLANE_NAMESPACE} deployment/etcd-data`
+----
++
+.. Scale up the etcd cluster by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:220]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+$ oc scale -n ${CONTROL_PLANE_NAMESPACE} statefulset/etcd --replicas=3
+----
++
+.. Wait for the etcd member pods to return and report as available by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:227]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+$ oc get -n ${CONTROL_PLANE_NAMESPACE} pods -l app=etcd -w
+----
++
+.. Scale up all etcd-writer deployments by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:234]: one dot

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
++
+[source,terminal]
+----
+$ oc scale deployment -n ${HOSTED_CLUSTER_NAMESPACE} --replicas=3 kube-apiserver openshift-apiserver openshift-oauth-apiserver
+----
++
+.. Restore reconciliation of the hosted cluster by entering the following command:
```
Comment (csrwng) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:241]: one dot

**Review by csrwng — COMMENTED** [review]
More comments

**Comment by heliubj18** [comment]
LGTM 
@jiezhao16 Could you please help review it and add lgtm. Thanks

**Comment by jiezhao16** [comment]
Some commands don't have prompt $, instead they have ' ' (e.g. step#5)
At step#15 namespace is wrong, it should be ${CONTROL_PLANE_NAMESPACE}
For Technology Preview section, should it state on AWS?

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+----
++
+... Make a local copy of the snapshot by entering the following command:
++
+[source,terminal]
+----
+$ oc cp -c etcd ${CONTROL_PLANE_NAMESPACE}/${ETCD_POD}:/var/lib/snapshot.db /tmp/etcd.snapshot.db
+----
++
+.. Make a copy of the snapshot database from etcd persistent storage.
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:86]:
```suggestion
.. Make a copy of the snapshot database from etcd persistent storage:
```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
++
+... Find a pod that is running and set its name as the value of `ETCD_POD: ETCD_POD=etcd-0`, then copy its snapshot database by entering the following command:
++
+[source,terminal]
+----
+$ oc cp -c etcd ${CONTROL_PLANE_NAMESPACE}/${ETCD_POD}:/var/lib/data/member/snap/db /tmp/etcd.snapshot.db
+----
++
+. Scale down the etcd statefulset by entering the following command:
++
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:103]:
```suggestion

```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+[source,terminal]
+----
+$ oc scale -n ${CONTROL_PLANE_NAMESPACE} statefulset/etcd --replicas=0
+----
++
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:108]:
```suggestion

```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 29]
```diff
@@ -0,0 +1,260 @@
...
+. Delete volumes for second and third members by entering the following command:
++
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:110]:
```suggestion

```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 30]
```diff
@@ -0,0 +1,260 @@
...
+[source,terminal]
+----
+$ oc delete -n ${CONTROL_PLANE_NAMESPACE} pvc/data-etcd-1 pvc/data-etcd-2
+----
++
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:115]:
```suggestion

```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 31]
```diff
@@ -0,0 +1,260 @@
...
+. Create a pod to access the first etcd member's data:
++
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:117]:
```suggestion

```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 32]
```diff
@@ -0,0 +1,260 @@
...
+          while true; do
+            sleep 1000
+          done
+      volumes:
+      - name: data
+        persistentVolumeClaim:
+          claimName: data-etcd-0
+EOF
+----
++
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:164]:
```suggestion

```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:1]: Overall, I try to avoid having more than 5-6 steps in one chunk, because it gets really hard to follow (there are 16 steps here). You can often use headings to break up the steps into smaller chunks. Like for this content, you could break them down into headings like:

- Set environment variables and scale down the API servers
- Take a snapshot of etcd
- Scale down the statefulset and delete volumes

etc.

Just something to think about for readability if you have some extra time to see if you could improve it.

**Review by kcarmichael08 — COMMENTED** [review]
Nice job adding this content! I had some things that I marked per the guidelines and an overall suggestion for the module.

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 33]
```diff
@@ -0,0 +1,260 @@
...
+[source,terminal]
+----
+$ oc patch -n ${HOSTED_CLUSTER_NAMESPACE} hostedclusters/${CLUSTER_NAME} -p '{"spec":{"pausedUntil":"true"}}' --type=merge
+----
+
+. Scale down API servers by entering the following commands:
++
+[source,terminal]
+----
+$ oc scale -n ${CONTROL_PLANE_NAMESPACE} deployment/kube-apiserver --replicas=0
```
[the remark this answers is not in the document]
Comment (obrown1205) [reply modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:41]: great idea! :)

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 34]
```diff
@@ -0,0 +1,260 @@
...
+.. Get the etcd image by entering the following command:
++
+[source,terminal]
+----
+ETCD_IMAGE=$(oc get -n ${CONTROL_PLANE_NAMESPACE} statefulset/etcd -o jsonpath='{ .spec.template.spec.containers[0].image }')
+----
++
+.. Create a pod that will allow access to etcd data:
++
+[source,terminal]
```
[the remark this answers is not in the document]
Comment (obrown1205) [reply modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:127]: good call :)

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 35]
```diff
@@ -0,0 +1,277 @@
...
+====
+This procedure requires API downtime.
+====
+
+.Prerequisites
+* The `oc` and `jq` binaries have been installed.
+
+.Procedure
+
+. First, set up your environment variables and scale down the API servers
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:21]:
```suggestion
. First, set up your environment variables and scale down the API servers:
```

File: modules/hosted-cluster-etcd-quorum-loss-recovery.adoc [hunk 36]
```diff
@@ -0,0 +1,277 @@
...
+... Scale down the `openshift-oauth-apiserver`:
++
+[source,terminal]
+----
+$ oc scale -n ${CONTROL_PLANE_NAMESPACE} deployment/openshift-oauth-apiserver --replicas=0
+----
+
+. Next, take a snapshot of etcd by using one of the following methods:
+
+.. Use a previously backed-up snapshot of etcd
```
Comment (kcarmichael08) [inline modules/hosted-cluster-etcd-quorum-loss-recovery.adoc:72]:
```suggestion
.. Use a previously backed-up snapshot of etcd.
```

**Review by kcarmichael08 — COMMENTED** [review]
Hi, sorry, a couple of very minor issues that I didn't see previously.
